integration-tests/tests/src/tests/observable.ts

```diff
@@ -0,0 +1,848 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2023 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+/**
+ * This suite aims to test observability of Realm, Realm.Object and all collection types.
+ * In particular, it aims to test:
+ * - Adding a listener will call the callback on change.
+ */
+
+import { expect } from "chai";
+
+import Realm, { CollectionChangeSet, DictionaryChangeSet, ObjectChangeSet } from "realm";
+
+import { openRealmBefore, openRealmBeforeEach } from "../hooks";
+import { createListenerStub } from "../utils/listener-stub";
+import { createPromiseHandle } from "../utils/promise-handle";
+
+type Observable = Realm | Realm.Object<any> | Realm.Results<any> | Realm.List<any> | Realm.Dictionary | Realm.Set<any>;
+
+function expectObservableMethods(observable: Observable) {
+  expect(observable.addListener).to.be.a("function");
+  expect(observable.removeListener).to.be.a("function");
+  expect(observable.removeAllListeners).to.be.a("function");
+}
+
+function expectRealmNotifications(
+  realm: Realm,
+  eventName: "change" | "schema" | "beforenotify",
+  expectedChangeSets: { schema?: Realm.CanonicalObjectSchema[] }[],
+) {
+  const handle = createPromiseHandle();
+  realm.addListener(
+    eventName,
+    createListenerStub(
+      handle,
+      ...expectedChangeSets.map(
+        (expectedChanges, c) => (realm: Realm, name: string, schema?: Realm.CanonicalObjectSchema[]) => {
+          expect(name).equals(eventName, `Realm change event #${c} name didn't match`);
+          expect(schema).deep.equals(expectedChanges.schema, `Realm change event #${c} schema didn't match`);
```

Comment:
The naming of `c` initially through me off a bit, thinking it may be misused as I didn't associate the name to an index. But what do you think of moving the part within `map(..)` to another function to speed up readability? Perhaps something like
```js
...expectedChangeSets.map(createRealmListenerCallback)
```
If so, then it could also apply to the other `expectXNotifications` functions.

Reply:
I tend do gravitate towards inline arrow functions for simple mappings / transformations like these, especially if they're not reused.